src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c

```diff
@@ -0,0 +1,458 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "pal_x509.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+
+#define INIT_LOCALS(name, ...) \
+    enum { __VA_ARGS__, count_##name }; \
+    jobject name[count_##name] = { 0 }; \
+
+#define RELEASE_LOCALS(name, env) \
+{ \
+    for (int i = 0; i < count_##name; ++i) \
+    { \
+        jobject local = name[i]; \
+        if (local != NULL) \
+            (*env)->DeleteLocalRef(env, local); \
+    } \
+} \
+
+#define VALUE_FAIL -1
+#define BUFFER_FAIL FAIL
+
+static int32_t PopulateByteArray(JNIEnv *env, jbyteArray source, uint8_t *dest, int32_t len);
+static int32_t PopulateString(JNIEnv *env, jstring source, char *dest, int32_t len);
+
+// Handles both DER and PEM formats
+jobject /*X509Certificate*/ AndroidCryptoNative_DecodeX509(const uint8_t *buf, int32_t len)
+{
+    assert(buf != NULL && len > 0);
+    JNIEnv* env = GetJNIEnv();
+
+    jobject ret = NULL;
+    INIT_LOCALS(loc, bytes, stream, certType, certFactory)
+
+    // byte[] bytes = new byte[] { ... }
+    // InputStream stream = new ByteArrayInputStream(bytes);
+    loc[bytes] = (*env)->NewByteArray(env, len);
+    (*env)->SetByteArrayRegion(env, loc[bytes], 0, len, (const jbyte*)buf);
+    loc[stream] = (*env)->NewObject(env, g_ByteArrayInputStreamClass, g_ByteArrayInputStreamCtor, loc[bytes]);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+
+    // CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+    // return (X509Certificate)certFactory.generateCertificate(stream);
+    loc[certType] = JSTRING("X.509");
+    loc[certFactory] = (*env)->CallStaticObjectMethod(env, g_CertFactoryClass, g_CertFactoryGetInstance, loc[certType]);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    ret = (*env)->CallObjectMethod(env, loc[certFactory], g_CertFactoryGenerateCertificate, loc[stream]);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    if (ret != NULL)
+        ret = ToGRef(env, ret);
+
+cleanup:
+    RELEASE_LOCALS(loc, env)
+    return ret;
+}
+
+// Encodes as DER format
+int32_t AndroidCryptoNative_EncodeX509(jobject /*X509Certificate*/ cert, uint8_t *buf, int32_t len)
+{
+    assert(cert != NULL);
+    JNIEnv *env = GetJNIEnv();
+    int32_t ret = BUFFER_FAIL;
+
+    // byte[] encoded = cert.getEncoded();
+    // return encoded.length
+    jbyteArray encoded =  (*env)->CallObjectMethod(env, cert, g_X509CertGetEncoded);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+    ret = PopulateByteArray(env, encoded, buf, len);
+
+cleanup:
+    (*env)->DeleteLocalRef(env, encoded);
+    return ret;
+}
+
+void CryptoNative_X509Destroy(jobject /*X509Certificate*/ cert)
+{
+    ReleaseGRef(GetJNIEnv(), cert);
+}
+
+jobject /*X509Certificate*/ CryptoNative_X509UpRef(jobject /*X509Certificate*/ cert)
+{
+    return AddGRef(GetJNIEnv(), cert);
```

Comment:
You don't have to rewire things in this PR, but I'd probably just put one `AndroidCrypto_ReleaseJObject` in the shim somewhere and have all of the typed safe handles ultimately call that in their ReleaseHandle, instead of making a different one per type.

In OpenSSL API each SOMETHING_free function has typed inputs, and does type-specific cleanup, so we needed the verbosity there.

Reply:
I'm planning on having two shim functions for lifetime (new global ref and delete global ref) and something like a SafeJObjecthandle that all the specifically typed handles could inherit from.

Started trying to do it in this PR at first, but decided to do that in another change because so much is still shared with Unix. (I'm planning on splitting off a lot more - stubbed out - in my next change).

Reply:
> I'm planning on splitting off a lot more - stubbed out - in my next change

Hooray! 🎉